lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -200,27 +200,39 @@ static bool hasConvertToMMATransisitiveUse(Operation *op, Attribute encoding) {
 static bool isLayoutAnchor(Operation *op) {
   if (isa<triton::LoadOp, triton::StoreOp>(op))
     return isExpensiveLoadOrStore(op);
-  if (isa<triton::ReshapeOp, triton::DotOp, triton::AtomicRMWOp,
-          triton::AtomicCASOp>(op))
+  if (isa<triton::DotOp, triton::AtomicRMWOp, triton::AtomicCASOp>(op))
     return true;
   return false;
 }
 
 void LayoutPropagation::initAnchorLayout() {
+  auto maybeAddAnchor = [&](Value v) {
+    if (auto tensorType = v.getType().dyn_cast<RankedTensorType>()) {
+      // Workaround, don't popagate MMA layout unless there is a convert
+      // back to mma further down to avoid generating reduction with MMA
+      // layout that may have lower performance.
+      // This can be improved with more aggressive backward propagation.
+      if (tensorType.getEncoding().isa<triton::gpu::NvidiaMmaEncodingAttr>() &&
+          v.getDefiningOp() &&
+          !hasConvertToMMATransisitiveUse(v.getDefiningOp(),
+                                          tensorType.getEncoding())) {
+        return;
+      }
+      layouts.insert({v, LayoutInfo(tensorType.getEncoding())});
+    }
+  };
+
+  // Consider function args as anchors.  This makes it easier to write tests --
+  // you can pass a tensor with an encoding as an arg, instead of explicitly
+  // calling tt.load.
+  for (auto arg : funcOp.getArguments()) {
+    maybeAddAnchor(arg);
+  }
+
   funcOp.walk([&](Operation *op) {
     if (isLayoutAnchor(op)) {
       for (auto result : op->getResults()) {
-        if (auto tensorType = result.getType().dyn_cast<RankedTensorType>()) {
-          // Workaround, don't popagate MMA layout unless there is a convert
-          // back to mma further down to avoid generating reduction with MMA
-          // layout that may have lower performance.
-          // This can be improved with more aggressive backward propagation.
-          if (tensorType.getEncoding()
-                  .isa<triton::gpu::NvidiaMmaEncodingAttr>() &&
-              !hasConvertToMMATransisitiveUse(op, tensorType.getEncoding()))
-            continue;
-          layouts.insert({result, LayoutInfo(tensorType.getEncoding())});
-        }
+        maybeAddAnchor(result);
```

Comment:
I think we still want to consider reshape with reorder as anchor. I expect that we will run into performance problem without that

Reply:
Sorry, this auto-merged right when your comment came in.

Why do we need to consider reshape with reorder as an anchor?  I thought it could handle any layout, so we don't care?

Reply:
> I thought it could handle any layout, so we don't care?

right, this is kind of an heuristic. It is because this op would stop the forward propagation (since the dst encoding can be anything) so if there are convert layout after they would not get folded. As a solution we propagate whatever layout was the dst of the reshape and the backward pass will revert and fold in the other direction if needed.

Reply:
ctx: https://github.com/openai/triton/pull/2268

Reply:
https://github.com/openai/triton/pull/3098